File: dsa-problems/Trees/solutions/TRE-016-robotics-balance-check-weight.py
import sys

# Increase recursion depth
sys.setrecursionlimit(200000)

def is_balanced_weighted(n: int, weight: list[int], left: list[int], right: list[int], W: int) -> bool:
    if n == 0:
        return True
        
    # Returns (height, total_weight, is_balanced)
    def dfs(u):
        if u == -1:
            return 0, 0, True
            
        h_l, w_l, bal_l = dfs(left[u])
        if not bal_l:
            return 0, 0, False
            
        h_r, w_r, bal_r = dfs(right[u])
        if not bal_r:
            return 0, 0, False
            
        h_bal = abs(h_l - h_r) <= 1
        w_bal = abs(w_l - w_r) <= W
        
        if h_bal and w_bal:
            return max(h_l, h_r) + 1, w_l + w_r + weight[u], True
        else:
            return 0, 0, False

    return dfs(0)[2]

def main():
    data = sys.stdin.read().strip().split()
    if not data:
        return
    idx = 0
    n = int(data[idx]); idx += 1
    weight = [0] * n
    left = [0] * n
    right = [0] * n
    for i in range(n):
        weight[i] = int(data[idx]); idx += 1
        left[i] = int(data[idx]); idx += 1
        right[i] = int(data[idx]); idx += 1
    W = int(data[idx]) if idx < len(data) else 0
    
    print("true" if is_balanced_weighted(n, weight, left, right, W) else "false")

if __name__ == "__main__":
    main()

File: dsa-problems/Trees/solutions/javascript/TRE-016-robotics-balance-check-weight.js
const readline = require("readline");

class Solution {
  isBalancedWeighted(n, weight, left, right, W) {
    if (n === 0) return true;

    const dfs = (u) => {
      if (u === -1) return { height: 0, weight: 0n, balanced: true };

      const l = dfs(left[u]);
      if (!l.balanced) return { height: 0, weight: 0n, balanced: false };

      const r = dfs(right[u]);
      if (!r.balanced) return { height: 0, weight: 0n, balanced: false };

      const hDiff = Math.abs(l.height - r.height);
      let wDiff = l.weight - r.weight;
      if (wDiff < 0n) wDiff = -wDiff;

      const hBal = hDiff <= 1;
      const wBal = wDiff <= BigInt(W);

      if (hBal && wBal) {
        return {
          height: Math.max(l.height, r.height) + 1,
          weight: l.weight + r.weight + BigInt(weight[u]),
          balanced: true,
        };
      } else {
        return { height: 0, weight: 0n, balanced: false };
      }
    };

    return dfs(0).balanced;
  }
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let data = [];
rl.on("line", (line) => data.push(...line.trim().split(/\s+/)));
rl.on("close", () => {
  if (data.length === 0) return;
  let idx = 0;
  const n = parseInt(data[idx++], 10);
  const weight = new Array(n);
  const left = new Array(n);
  const right = new Array(n);
  for (let i = 0; i < n; i++) {
    weight[i] = parseInt(data[idx++], 10);
    left[i] = parseInt(data[idx++], 10);
    right[i] = parseInt(data[idx++], 10);
  }
  const W = idx < data.length ? parseInt(data[idx], 10) : 0;

  const solution = new Solution();
  console.log(solution.isBalancedWeighted(n, weight, left, right, W) ? "true" : "false");
});
